src/main/java/com/mongodb/jdbc/MongoSQLDatabaseMetaData.java

```diff
@@ -161,130 +237,766 @@ public ResultSet getTableTypes() throws SQLException {
+                new BsonElement(GRANTEE, new BsonString("")),
+                new BsonElement(PRIVILEGE, new BsonString("SELECT")),
+                new BsonElement(IS_GRANTABLE, BsonNull.VALUE));
+    }
+
+    // Helper for getting table data for all tables from a specific database. Used by
+    // getTables and getTablePrivileges. The caller specifies how to serialize the table
+    // info into BSON documents for the result set.
+    private Stream<BsonDocument> getTableDataFromDB(
+            String dbName,
+            Pattern tableNamePatternRE,
+            List<String> types,
+            BiFunction<String, MongoListCollectionsResult, BsonDocument> bsonSerializer) {
+
+        return this.getTableDataFromDB(
+                        dbName,
+                        res ->
+                                tableNamePatternRE.matcher(res.name).matches()
+                                        && (types == null
+                                                || types.contains(res.type.toLowerCase())))
+                .map(res -> bsonSerializer.apply(dbName, res));
+    }
+
+    private List<String> toTableTypeList(String[] types) {
+        List<String> l = null;
+        if (types != null) {
+            l = Arrays.asList(types);
+            l.replaceAll(String::toLowerCase);
+        }
+        return l;
+    }
+
     @Override
     public ResultSet getTables(
-            String catalog, String schemaPattern, String tableNamePattern, String types[])
+            String catalog, String schemaPattern, String tableNamePattern, String[] types)
             throws SQLException {
-        throw new SQLFeatureNotSupportedException("TODO");
+        MongoJsonSchema botSchema =
+                createBottomSchema(
+                        new Pair<>(TABLE_CAT, BsonTypeInfo.STRING_TYPE_NAME),
+                        new Pair<>(TABLE_SCHEM, BsonTypeInfo.STRING_TYPE_NAME),
+                        new Pair<>(TABLE_NAME, BsonTypeInfo.STRING_TYPE_NAME),
+                        new Pair<>(TABLE_TYPE, BsonTypeInfo.STRING_TYPE_NAME),
+                        new Pair<>(REMARKS, BsonTypeInfo.STRING_TYPE_NAME),
+                        new Pair<>(TYPE_CAT, BsonTypeInfo.STRING_TYPE_NAME),
+                        new Pair<>(TYPE_SCHEM, BsonTypeInfo.STRING_TYPE_NAME),
+                        new Pair<>(TYPE_NAME, BsonTypeInfo.STRING_TYPE_NAME),
+                        new Pair<>(SELF_REFERENCING_COL_NAME, BsonTypeInfo.STRING_TYPE_NAME),
+                        new Pair<>(REF_GENERATION, BsonTypeInfo.STRING_TYPE_NAME));
+
+        // Note: JDBC has Catalogs, Schemas, and Tables: they are three levels of organization.
+        // MongoDB only has Databases (Catalogs) and Collections (Tables), so we ignore the
+        // schemaPattern argument.
+        Pattern tableNamePatternRE = toJavaPattern(tableNamePattern);
+        List<String> typesList = toTableTypeList(types);
+
+        Stream<BsonDocument> docs;
+        if (catalog == null) {
```

Comment:
Oh thanks for trying that. My expectation was that `catalog = ""` wouldn't need special handling in any of these methods. I figured doing `this.conn.getDatabase("").command(...).stream()...` would just naturally result in empty result set since there would never be a database with that name.

I'll update these methods to explicitly handle the empty string case.

Side question for my own understanding: where was that IllegalArgumentException thrown? was it thrown by the java driver? something in our code? somewhere else?

Reply:
It is thrown in the java driver, the check for the empty string is [here](https://github.com/mongodb/mongo-java-driver/blob/a04a96c3f1afe044bc4396b81244608a95c6494d/driver-core/src/main/com/mongodb/MongoNamespace.java#L59).

The stack trace is:
```
state should be: databaseName is not empty
java.lang.IllegalArgumentException: state should be: databaseName is not empty
	at com.mongodb.assertions.Assertions.isTrueArgument(Assertions.java:101)
	at com.mongodb.MongoNamespace.checkDatabaseNameValidity(MongoNamespace.java:59)
	at com.mongodb.client.internal.MongoDatabaseImpl.<init>(MongoDatabaseImpl.java:74)
	at com.mongodb.client.internal.MongoClientImpl.getDatabase(MongoClientImpl.java:80)
	at com.mongodb.jdbc.MongoConnection.getDatabase(MongoConnection.java:97)
	at com.mongodb.jdbc.MongoSQLDatabaseMetaData.getTableDataFromDB(MongoSQLDatabaseMetaData.java:251)
	at com.mongodb.jdbc.MongoSQLDatabaseMetaData.getTableDataFromDB(MongoSQLDatabaseMetaData.java:305)
	at com.mongodb.jdbc.MongoSQLDatabaseMetaData.getTables(MongoSQLDatabaseMetaData.java:359)
```

Reply:
Thanks!